child_process: add 'overlapped' stdio flag

tarruda opened the pull request:
The 'overlapped' value sets the UV_OVERLAPPED_PIPE libuv flag in the child process stdio.

Fixes: https://github.com/nodejs/node/issues/29238

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests added
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

mscdex commented:
Why not just always set `UV_OVERLAPPED_PIPE` for `'pipe'` instead? Are there any drawbacks/incompatibilities with that?

tarruda commented:
> Why not just always set UV_OVERLAPPED_PIPE for 'pipe' instead? Are there any drawbacks/incompatibilities with that?

Yes, it can possibly break existing programs that run child process that are not prepared for overlapped I/O: https://github.com/libuv/libuv/issues/95#issuecomment-68413692

mscdex commented:
Ok, perhaps we can just shorten the name to `'overlapped'` instead?

tarruda commented:
> Ok, perhaps we can just shorten the name to 'overlapped' instead?

Added a fixup. If this is acceptable I will squash and edit the commit title.

mscdex commented:
Can you also add a test for this?

tarruda commented:
> Can you also add a test for this?

Do you have any suggestions on how I should proceed to testing this?

The new option simply activates the `UV_OVERLAPPED_PIPE` libuv flag, is there any existing test infrastructure to verify that the correct flags were passed by the new option? I couldn't find any reference to `UV_CREATE_PIPE` in the tests, so I imagine the answer is "no". 

I suppose I could just add a functional test that verifies overlapped I/O on the child process, but that would probably be duplicating one of the libuv tests. 

Also, the code path should be close to the one followed by the`'pipe'` value. Would it be enough to copy one of the `'pipe'` tests but pass `'overlapped'` instead?

mscdex commented:
At the very least there should be a test that specifying `'overlapped'` (via `stdio: 'overlapped'` or manually expanded out to `stdio: ['overlapped', 'overlapped', 'overlapped']`) does not throw an exception (since the code currently throws on unsupported modes).

tarruda commented:
> At the very least there should be a test that specifying 'overlapped' (via stdio: 'overlapped' or manually expanded out to stdio: ['overlapped', 'overlapped', 'overlapped']) does not throw an exception (since the code currently throws on unsupported modes).

:+1: 

I also want to add a test that runs a C++  windows program that will not work if the stdio handles are not overlapped pipes.

tarruda commented:
I've added "overlapped-checker", a program which is used to verify the `FILE_FLAG_OVERLAPPED` status from the child. The new test uses this program to verify that the flag is correctly passed.

tarruda commented:
PR ready for final review.

Trott commented:
@nodejs/testing @nodejs/build-files @nodejs/gyp Thoughts on the separate `overlapped-checker` program here used for testing?

Trott commented:
Also: Welcome @tarruda and thanks for the pull request! These kinds of things can often attract a lot of comments, so thanks in advance for your patience.

bnoordhuis commented on test/overlapped-checker/main.c:
```diff
@@ -0,0 +1,23 @@
+#include <stdlib.h>
+
+#include "../../deps/uv/include/uv.h"
+#include "../../deps/uv/src/uv-common.h"
```
This is to get the definition of `UV_HANDLE_NON_OVERLAPPED_PIPE`, right?

Depending on internal libuv headers is bad because there's no guarantee they stay usable (or even present.) This could start failing at the next libuv upgrade.

tarruda replied:
I've looked into how libuv gets the overlapped information (the `pNtQueryInformationFile` function) and it seems to be using an internal windows API with manually defined structs in `winapi.h`. I could copy the definitions to avoid depending on libuv internal headers, but duplicating a bunch of doesn't seem much better.

Any thoughts?

bnoordhuis replied:
`NtQueryInformationFile()` is not an internal API. The reason libuv defines it manually is because its definition comes from the DDK (driver development kit) which isn't normally installed. The API itself is always available though.

My preference would be to obtain the overlapped flag from the kernel. You can probably get away with a minimal definition of the relevant structs and enums. You'll need `IO_STATUS_BLOCK` and `FILE_MODE_INFORMATION`; their members are basic types.

`FileModeInformation == 16`, you could just pass that as a literal `/* FileModeInformation */ 16`.

You can steal the actual logic from libuv. :-)

https://github.com/libuv/libuv/blob/549f52c4672a4d8198f096efa513a96b20e2f843/src/win/pipe.c#L297-L310

tarruda replied:
Yesterday I rewrote the test script. Now it is a bit more complex but does check if `FILE_FLAG_OVERLAPPED` is set implicitly (the test program deadlocks if not) without depending on libuv internal headers. Can you review it again and let me know if it is an acceptable alternative to using `NtQueryInformationFile` to check the file flags?

bnoordhuis commented on test/overlapped-checker/main.c:
```diff
@@ -0,0 +1,23 @@
+#include "../../deps/uv/include/uv.h"
+#include "../../deps/uv/src/uv-common.h"
+
+int main() {
```
`int main(void) {` - `int main() {` is not a legal definition of main (it is in C++ but not in C.)

The function also needs a `return 0;` statement.

bnoordhuis commented on test/overlapped-checker/main.c:
```diff
@@ -0,0 +1,23 @@
+#include "../../deps/uv/src/uv-common.h"
+
+int main() {
+  uv_loop_t *loop = uv_default_loop();
```
Style nit: `uv_loop_t* loop = ...` (star leans left, ditto on line 20.)

bnoordhuis commented on test/overlapped-checker/main.c:
```diff
@@ -0,0 +1,23 @@
+      abort();
+    }
+
+    int non_overlapped_flag = !!(pipe.flags & UV_HANDLE_NON_OVERLAPPED_PIPE);
```
This is unreliable. Libuv's UNIX pipe code could add a flag with the same value as `UV_HANDLE_NON_OVERLAPPED_PIPE` (which is Windows-only) and then this code starts printing different numbers.

bnoordhuis commented on test/overlapped-checker/main.c:
```diff
@@ -0,0 +1,23 @@
+
+    int non_overlapped_flag = !!(pipe.flags & UV_HANDLE_NON_OVERLAPPED_PIPE);
+
+    fprintf(stderr, "%d", non_overlapped_flag);
```
You should fflush() or print a newline on exit because stderr might be in line-buffered mode, see `man stdbuf`.

tarruda commented:
@bnoordhuis addressed your comments except the usage of `UV_HANDLE_NON_OVERLAPPED_PIPE`. Do you have a suggestion on how to handle this without depending on an internal libuv header?

Trott commented on doc/api/child_process.md:
```diff
@@ -613,7 +614,13 @@ pipes between the parent and child. The value is one of the following:
-2. `'ipc'` - Create an IPC channel for passing messages/file descriptors
+2. `'overlapped'` - Same as `'pipe'` except that the `FILE_FLAG_OVERLAPPED` flag
+   is set on the handle. This is necessary for overlapped I/O on the child
+   process's stdio handles, read the
```
```suggestion
   process's stdio handles. See the
```

tarruda replied:
Addressed

bnoordhuis commented on test/overlapped-checker/main_unix.c:
```diff
@@ -0,0 +1,35 @@
+#include <unistd.h>
+
+static size_t r(char *buf, size_t buf_size) {
+  ssize_t read_count = read(0, buf, buf_size);
```
This needs to be something like this:
```c
// #include <errno.h>
ssize_t read_count;
do
  read_count = read(0, buf, buf_size);
while (read_count == -1 && errno == EINTR);
```
Ditto for the `write()` call. Reads and writes can be interrupted by signals. That's benign but needs to be handled.

bnoordhuis commented on test/overlapped-checker/main_unix.c:
```diff
@@ -0,0 +1,35 @@
+  return (size_t)read_count;
+}
+
+static void w(const char* buf, size_t count) {
```
Style: `const char* buf` (star leans left, ditto online 5.)

bnoordhuis commented on test/overlapped-checker/main_unix.c:
```diff
@@ -0,0 +1,35 @@
+}
+
+static void w(const char* buf, size_t count) {
+  ssize_t write_count = write(1, buf, count);
```
You should also be prepared for partial writes, besides EINTR.

bnoordhuis commented on test/overlapped-checker/main_unix.c:
```diff
@@ -0,0 +1,35 @@
+  while (1) {
+    char buf[256];
+    size_t read_count = r(buf, sizeof(buf));
+    if (!strncmp(buf, "exit", read_count)) {
```
This doesn't look Obviously Correct (because of partial reads) although it probably is (but it took me a while to figure out.)

Reads and writes on pipes are guaranteed to be atomic up til `PIPE_BUF` bytes (guaranteed to be at least 512 bytes but possibly much more) but it's not clear from reading this code whether the input is or isn't < `PIPE_BUF`.

Reading through the JS test I'd say it is but it's not super obvious.

tarruda replied:
Should I do anything to make it more clear? Maybe add a remark regarding reads up to `PIPE_BUF` being atomic?

bnoordhuis replied:
An explaining comment would be helpful, yes.

bnoordhuis commented on test/overlapped-checker/main_win.c:
```diff
@@ -0,0 +1,67 @@
+#include <windows.h>
+
+char buf[256];
+DWORD read_count = 0, write_count = 0;
```
Assignment isn't really necessary, they're initialized to 0 anyway because they're in the bss section.

Style note: prefer one definition per line:
```c
DWORD read_count;
DWORD write_count;
```

bnoordhuis commented on test/overlapped-checker/main_win.c:
```diff
@@ -0,0 +1,67 @@
+char buf[256];
+DWORD read_count = 0, write_count = 0;
+HANDLE stdin_h;
+OVERLAPPED stdin_o;
```
These should all be `static`.

bnoordhuis commented on test/overlapped-checker/main_win.c:
```diff
@@ -0,0 +1,67 @@
+HANDLE stdin_h;
+OVERLAPPED stdin_o;
+
+static void die(const char *buf) {
```
Style: `const char* buf`. Also on line 28.

bnoordhuis commented on test/overlapped-checker/main_win.c:
```diff
@@ -0,0 +1,67 @@
+
+int main(void) {
+  HANDLE event = CreateEvent(NULL, FALSE, FALSE, NULL);
+  if (!event) {
```
Style: mild preference for `if (event == NULL) {` (no type coercion)

bnoordhuis commented on test/parallel/test-child-process-stdio-overlapped.js:
```diff
@@ -0,0 +1,79 @@
+//   cycle until the number reaches 200, at which point we send the "exit"
+//   string, which causes the test program to exit.
+// - Extra assertion: Every time the test program writes a string to its stdout,
+//   it will write the number of bytes written to stderr.
```
Can you make it explicit here that the helper program is supposed to hang if overlapped I/O is somehow not working?

tarruda replied:
Above there's the following comment:

```
// - Verify that overlapped I/O works on windows. The test program will deadlock
//   if stdin doesn't have the FILE_FLAG_OVERLAPPED flag set on startup (see
//   test/overlapped-checker/main_win.c for more details).
```

Is that enough?

bnoordhuis commented on test/parallel/test-child-process-stdio-overlapped.js:
```diff
@@ -0,0 +1,79 @@
+  stderr += s;
+});
+
+child.stderr.on('end', () => {
```
Forgot to mention: wrap the callback in a `common.mustCall()` here and on line 76.

gireeshpunathil commented:
ping @tarruda

tarruda commented:
I've addressed most comments, @bnoordhuis can you review?

bnoordhuis commented on test/overlapped-checker/main_win.c:
```diff
@@ -0,0 +1,68 @@
+  }
+
+  stdin_h = GetStdHandle(STD_INPUT_HANDLE);
+  memset(&stdin_o, 0, sizeof(stdin_o));
```
Superfluous memset() since `stdin_o` is located in static storage. (Also, I think you need to include `<string.h>` but maybe `<windows.h>` also pulls in its definition.)

tarruda commented:
> Thanks, LGTM. I'll kick off the CI.

I'm still addressing one last comment (the GetLastError one)

tarruda commented:
Should I squash/rebase?

tarruda commented on test/overlapped-checker/main_win.c:
```diff
@@ -0,0 +1,80 @@
+          sizeof(msgbuf),
+          NULL);
+      snprintf(emsg, sizeof(emsg), "WaitForSingleObject failed. Error %d (%s)",
+          ecode, msgbuf);
```
@bnoordhuis can you check the usage of FormatMessage? I get a bit confused about the typedefs used in Windows API.

tarruda commented:
rebased and squashed fixup commits

tarruda commented:
I'm going to rebase but I see `.travis.yml` is no longer present. Can someone direct me to the current CI script?

bnoordhuis commented:
@tarruda You can just `git rm` it. We don't use Travis CI anymore, it's all GH Actions now.

tarruda commented:
@bnoordhuis it seems some of the builds are failing. I tried navigating the details to determine what caused the failure but got a bit lost, can you help?

tarruda commented:
rebased

tarruda commented:
All tests are passing, do I need to do anything else?

tarruda commented:
fixed failing tests and rebased

tarruda commented:
@richardlau squashed the fixup commit

tarruda commented:
ping @Fishrock123

Fishrock123 commented:
I don’t want to review this? Please find someone else for that, I don’t really work on Node anymore.

tarruda commented:
> Sorry I don’t really want to review this? Please find someone else for that, I don’t really work on Node anymore

Sorry, I didn't knew. It is just that merging is blocked awaiting for your review.

aduh95 reviewed (changes requested):
Can you add an entry in the `changes` list in `child_process.md` please?

```diff
--- a/doc/api/child_process.md
+++ b/doc/api/child_process.md
@@ -614,6 +614,9 @@ subprocess.unref();
 <!-- YAML
 added: v0.7.10
 changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/29412
+    description: Added the `overlapped` stdio flag.
   - version: v3.3.1
     pr-url: https://github.com/nodejs/node/pull/2727
     description: The value `0` is now accepted as a file descriptor.
```

Otherwise LGTM if the CI is green.

tarruda commented:
Rebased and added doc change quested by @aduh95

aduh95 commented on doc/api/child_process.md:
```diff
@@ -642,7 +646,13 @@ pipes between the parent and child. The value is one of the following:
    created for fds 0, 1, and 2 are also available as [`subprocess.stdin`][],
    [`subprocess.stdout`][] and [`subprocess.stderr`][], respectively.
-2. `'ipc'`: Create an IPC channel for passing messages/file descriptors
+2. `'overlapped'`: Same as `'pipe'` except that the `FILE_FLAG_OVERLAPPED` flag
```
```suggestion
1. `'overlapped'`: Same as `'pipe'` except that the `FILE_FLAG_OVERLAPPED` flag
```

aduh95 commented on doc/api/child_process.md:
```diff
@@ -642,7 +646,13 @@ pipes between the parent and child. The value is one of the following:
+   [docs](https://docs.microsoft.com/en-us/windows/win32/fileio/synchronous-and-asynchronous-i-o)
+   for more details. This is exactly the same as `'pipe'` on non-Windows
+   systems.
+3. `'ipc'`: Create an IPC channel for passing messages/file descriptors
```
```suggestion
1. `'ipc'`: Create an IPC channel for passing messages/file descriptors
```

aduh95 commented on doc/api/child_process.md:
```diff
@@ -653,25 +663,25 @@ pipes between the parent and child. The value is one of the following:
    or using the IPC channel with a child process that is not a Node.js instance
    is not supported.
-3. `'ignore'`: Instructs Node.js to ignore the fd in the child. While Node.js
+4. `'ignore'`: Instructs Node.js to ignore the fd in the child. While Node.js
```
```suggestion
1. `'ignore'`: Instructs Node.js to ignore the fd in the child. While Node.js
```

aduh95 commented on doc/api/child_process.md:
```diff
@@ -653,25 +663,25 @@ pipes between the parent and child. The value is one of the following:
    to `'ignore'` will cause Node.js to open `/dev/null` and attach it to the
    child's fd.
-4. `'inherit'`: Pass through the corresponding stdio stream to/from the
+5. `'inherit'`: Pass through the corresponding stdio stream to/from the
```
```suggestion
1. `'inherit'`: Pass through the corresponding stdio stream to/from the
```

aduh95 commented on doc/api/child_process.md:
```diff
@@ -653,25 +663,25 @@ pipes between the parent and child. The value is one of the following:
    `process.stdin`, `process.stdout`, and `process.stderr`, respectively. In
    any other position, equivalent to `'ignore'`.
-5. {Stream} object: Share a readable or writable stream that refers to a tty,
+6. {Stream} object: Share a readable or writable stream that refers to a tty,
```
```suggestion
1. {Stream} object: Share a readable or writable stream that refers to a tty,
```

aduh95 commented on doc/api/child_process.md:
```diff
@@ -653,25 +663,25 @@ pipes between the parent and child. The value is one of the following:
    underlying descriptor (file streams do not until the `'open'` event has
    occurred).
-6. Positive integer: The integer value is interpreted as a file descriptor
+7. Positive integer: The integer value is interpreted as a file descriptor
```
```suggestion
1. Positive integer: The integer value is interpreted as a file descriptor
```

aduh95 commented on doc/api/child_process.md:
```diff
@@ -653,25 +663,25 @@ pipes between the parent and child. The value is one of the following:
    process, similar to how {Stream} objects can be shared. Passing sockets
    is not supported on Windows.
-7. `null`, `undefined`: Use default value. For stdio fds 0, 1, and 2 (in other
+8. `null`, `undefined`: Use default value. For stdio fds 0, 1, and 2 (in other
```
```suggestion
1. `null`, `undefined`: Use default value. For stdio fds 0, 1, and 2 (in other
```

aduh95 reviewed (approved):
nit: markdown allows to use `1.` for every items in an ordered list. We could use that here, so if another item was added in the future, it would create cleaner diff. Either way, I'm OK landing this if CI is green.

tarruda commented:
@aduh95 This is the failure I'm getting:

```
Error: spawn C:\workspace\node-test-binary-windows-js-suites\node\out\Release\overlapped-checker.exe ENOENT
```

It is not finding the overlapped-checker program that tests the new flag. Do you have any idea how to fix this?

aduh95 reviewed (dismissed):
I can't compile the tests on my local machine, I'm getting the same error as macOS CI: https://github.com/nodejs/node/pull/29412/checks?check_run_id=1371388101

tarruda commented:
@aduh95 Can you try again? I have fixed the main_unix error, so you should be able to test on mac.

aduh95 reviewed (commented):
Works on my machine, thanks! Let's see if the CI passes now.

tarruda commented:
For some reason some of the builds cannot find the overlapped-checker program in Release

aduh95 commented:
> For some reason some of the builds cannot find the overlapped-checker program in Release

Yeah it seams so… I don't know gyp enough to understand the error though. @nodejs/gyp can you help decipher the CI errors?

tarruda commented:
@aduh95 I have rebased and it seems CI is now passing

tarruda commented:
@aduh95 I did some investigation and found out why overlapped-checker.exe is missing in the CI.

It seems the CI scripts are in https://github.com/nodejs/build which creates a binary.tar.gz file with all binaries for running in some of the jobs. I would need to modify https://github.com/nodejs/build/blob/master/jenkins/scripts/windows/node-compile-windows.cmd in order to include overlapped-checker.exe. 

Unfortunately I don't know how to make this PR's CI use my fork of https://github.com/nodejs/build. Do you have any hints for me?

tarruda commented:
I've temporarily edited the only  https://github.com/nodejs/build I could find to point to my [fork](https://github.com/tarruda/build/tree/expose-stdio-overlapped-pipe). Let's see if it is enough to make the CI pass...

tarruda commented:
@aduh95 can you request-ci?

aduh95 commented:
> @aduh95 can you request-ci?

@tarruda for security reasons, the CI is configured to pull code from repos on @nodejs org only. I'm afraid the only thing left to do for us is to wait for https://github.com/nodejs/build/pull/2495 to be reviewed, to land and to be deployed – unless of course you are able to find a workaround in the mean time. I understand this may be frustrating, but I hope you can understand we cannot allow any code to be deployed on our CI setup before it gets audited by folks who can check that it won't break anything.

BTW the change you made to the `Makefile` should be reverted, it would affect coverage builds, not the CI itself ;)

tarruda commented:
> I hope you can understand we cannot allow any code to be deployed on our CI setup before it gets audited by folks who can check that it won't break anything.

Totally understandable, removed the Makefile change.

> unless of course you are able to find a workaround in the mean time

I've done some investigation but unfortunately didn't process much beyond the Makefile change, it was the only connection between this repo and nodejs/build I could find. 

TBH I have no idea how nodejs CI works right now, not too mention I'm not even familiar with github actions (when this PR was first created, nodejs was still using travis). I'm unlikely to find any workaround without the help of CI team, though I'm hopeful  nodejs/build#2495 will fix the error after it is merged.

tarruda commented:
@aduh95 I've rebased, can we get another request-ci?

tarruda commented:
@aduh95 I navigated a bit more through the scripts in the build repository and found that the failing ci job (`node-test-binary-windows-js-suites`) eventually calls `vcbuild.bat` (in this repo) with `nobuild` argument. According to @joaocgreis, it must be built in this case:

> When using the "Resume build" button in Jenkins, it tries to use the results from old jobs that passed. This is very useful in some situations, but in this case we need a proper full Rebuild

What I did was handle the `test-ci-js` target in `vcbuild.bat` by resetting `nobuild` in order to always build overlapped-checker.exe.  I'm not sure if this is the correct path, but hopefully the build will pass now. Kindly `request-ci` so we can verify.

tarruda commented:
Following @joaocgreis suggestion on how to make the test find overlapped-checker.exe:

> To move forward, if I'm right about this, the easiest way is to find overlapped-checker relative to the executable, like what is done for openssl-cli in

tarruda commented:
it seems @joaocgreis's suggestion worked, rebasing on master a squashing the fixup

aduh95 reviewed (approved):
RSLGTM because Ben approved it already, but I'd feel more confident if we can get more approvals from other collaborators before landing.

tarruda commented:
@aduh95 Not sure if it helps, but I can try to explain the PR to make you more confident about merging.

The "overlapped" stdio option is simply exposing libuv's `UV_OVERLAPPED_PIPE`, which itself is a wrapper around `FILE_FLAG_OVERLAPPED` win32 flag. All this flag does it enable non-blocking I/O on child process's std handles and has no effect on Unix systems (it is the same as "pipe" on Unix).

Most of the code in the PR is dedicated to testing that the flag is passed correctly (AFAIR, libuv doesn't have any test covering it). The way the test works is that it will hang if the `FILE_FLAG_OVERLAPPED` flag is not passed to the child process. If you want, I can push a temporary commit that will modify the test to pass `pipe` instead of `overlapped` and we can see the test hang.

tarruda commented:
@aduh95 I've added a temporary commit that removes the "overlapped" flag. If you request-ci, we should see windows test hang

aduh95 commented on test/parallel/test-child-process-stdio-overlapped.js:
```diff
@@ -0,0 +1,80 @@
+  // - 150 (3 bytes written)
+  // - 200 (3 bytes written)
+  assert.strictEqual(stderr, '12333');
+  clearTimeout(timer);
```
```suggestion
```

aduh95 reviewed (commented):
Thanks for the explanation, it's clear to me now.

Can you remove the `timer` from the test please? There is a default timeout set by the `test.py` script, so the timeout can be customized by the test runner.

https://github.com/nodejs/node/blob/e57d8af7e2890b03d865c6edac279e234c43b0ae/tools/test.py#L1333-L1334

aduh95 commented:
@tarruda Do you want to revert 45e83ab to have this ready to land?

tarruda commented:
> @tarruda Do you want to revert 45e83ab to have this ready to land?

Removed it and added a fixup with the timeout removal on the tests

tarruda commented:
@aduh95 `node-test-linux-linked-debug ` failed, but it seems unrelated to this PR. Logs show that the ld process was killed during linking of `cctest`

aduh95 commented:
> @aduh95 `node-test-linux-linked-debug ` failed, but it seems unrelated to this PR. Logs show that the ld process was killed during linking of `cctest`

Probably unrelated indeed, it happens sometimes. I've kicked off another CI to make sure.

Benchmark CI: https://ci.nodejs.org/job/node-test-pull-request/35168/

aduh95 commented:
Landed in 25b21e478e82 🎉

tarruda commented:
Thanks @aduh95 @joaocgreis @bnoordhuis for the help with this PR
